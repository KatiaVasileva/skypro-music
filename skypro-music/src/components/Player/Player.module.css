.bar {
    position: absolute;
    bottom: 0;
    left: 0;
    width: 100%;
    background: rgba(28, 28, 28, 0.5);
}

.content {
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-orient: vertical;
    -webkit-box-direction: normal;
    -ms-flex-direction: column;
    flex-direction: column;
}

.playerBlock {
    height: 73px;
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-orient: horizontal;
    -webkit-box-direction: normal;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-box-pack: justify;
    -ms-flex-pack: justify;
    justify-content: space-between;
    background-color: #1C1C1C;
}

.player {
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-orient: horizontal;
    -webkit-box-direction: normal;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    -webkit-box-pack: start;
    -ms-flex-pack: start;
    justify-content: flex-start;
}

.controls {
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-orient: horizontal;
    -webkit-box-direction: normal;
    -ms-flex-direction: row;
    flex-direction: row;
    padding: 0 27px 0 31px;
    width: 300px;
}

.buttonPrev, .buttonPlay, .buttonPause, .buttonNext, .buttonRepeat, .buttonShuffle {
    padding: 5px;
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
}

.buttonPlay, .buttonPause, .buttonPrev, .buttonNext {
    width: 23px;
    margin-right: 23px;
    cursor: pointer;
}

.buttonNext {
    margin-right: 28px;
    fill: #a53939;
}

.buttonPrevSvg, .buttonNextSvg {
    width: 15px;
    height: 14px;
    fill: inherit;
    stroke: #d9d9d9;
}

.buttonPlaySvg {
    width: 22px;
    height: 20px;
    fill: #d9d9d9;
}

.buttonPauseSvg {
    width: 15px;
    height: 19px;
    fill: #d9d9d9;
}

.buttonRepeat {
    margin-right: 24px;
}

.buttonRepeatSvg {
    width: 18px;
    height: 12px;
    fill: transparent;
    stroke: #696969;

    &:hover {
        fill: transparent;
        stroke: #acacac;
        cursor: pointer;
    }
}

.buttonRepeatSvgActive {
    width: 18px;
    height: 12px;
    fill: #696969;
    stroke: #ffffff;
    cursor: pointer;
}

.buttonShuffle {
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
}

.buttonShuffleSvg {
    width: 19px;
    height: 12px;
    fill: transparent;
    stroke: #696969;

    &:hover {
        fill: transparent;
        stroke: #acacac;
        cursor: pointer;
    }
}

.buttonShuffleSvgActive {
    width: 19px;
    height: 12px;
    fill: #696969;
    stroke: #ffffff;
    cursor: pointer;
}

.trackPlay {
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-orient: horizontal;
    -webkit-box-direction: normal;
    -ms-flex-direction: row;
    flex-direction: row;
}

.trackPlayContainer {
    width: 500px;
    display: -ms-grid;
    display: grid;
    -ms-grid-columns: auto 1fr;
    grid-template-columns: auto 1fr;
    grid-template-areas: "image author""image album";
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
}

.trackPlayImage {
    width: 51px;
    height: 51px;
    background-color: #313131;
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    -webkit-box-pack: center;
    -ms-flex-pack: center;
    justify-content: center;
    margin-right: 12px;
    grid-row: 1;
    -ms-grid-row-span: 2;
    grid-column: 1;
    grid-area: image;
}

.trackPlayImageSvg {
    width: 18px;
    height: 17px;
    fill: transparent;
    stroke: #4e4e4e;
}

.trackPlayName {
    grid-row: 1;
    grid-column: 2;
    grid-area: author;
    min-width: 260px;
}

.trackPlayNameLink {
    font-style: normal;
    font-weight: 400;
    font-size: 16px;
    line-height: 24px;
    color: #ffffff;
    white-space: nowrap;
}

.trackPlayAuthor {
    grid-row: 2;
    grid-column: 2;
    grid-area: album;
    min-width: 260px;
}

.trackPlayAuthorLink {
    font-style: normal;
    font-weight: 400;
    font-size: 13px;
    line-height: 24px;
    color: #ffffff;
}

.trackTime {
    width: 150px;
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
}

.trackTimeFigures {
    width: 32px;
    margin: 5px;
}

.trackPlayLikeDis {
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-orient: horizontal;
    -webkit-box-direction: normal;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    margin-left: 26%;
}

.trackPlayLike, .trackPlayDislike {
    padding: 5px;

    &:active {
        fill: #696969;
        stroke: #ffffff;
        cursor: pointer;
    }
}

.trackPlayLikeSvg, .trackPlayLikeSvgActive {
    width: 14px;
    height: 12px;
    fill: transparent;
    stroke: #696969;

    /* &:hover {
        fill: transparent;
        stroke: #acacac;
        cursor: pointer;
    }

    &:active {
        fill: transparent;
        stroke: #ffffff;
        cursor: pointer;
    } */
}

.trackPlayLikeSvgActive {
    fill: #b672ff;
}

/* .trackPlayDislike {
    margin-left: 28.5px;

    &:active {
        fill: #696969;
        stroke: #ffffff;
        cursor: pointer;
    }
} */

.trackPlayDislikeSvg {
    width: 14.34px;
    height: 13px;
    fill: transparent;
    stroke: #696969;

    &:hover {
        fill: transparent;
        stroke: #acacac;
        cursor: pointer;
    }

    &:active {
        fill: transparent;
        stroke: #ffffff;
        cursor: pointer;
    }
}

.volumeBlock {
    width: auto;
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    padding: 0 92px 0 0;
}

.volumeContent {
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-orient: horizontal;
    -webkit-box-direction: normal;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    -webkit-box-pack: end;
    -ms-flex-pack: end;
    justify-content: end;
}

.volumeImage {
    width: 13px;
    height: 18px;
    margin-right: 17px;
}

.volumeSvg {
    width: 13px;
    height: 18px;
    fill: transparent;
}

.volumeProgress {
    width: 109px;
    cursor: pointer;
}

.volumeProgressLine {
    width: 109px;
    cursor: pointer;
}