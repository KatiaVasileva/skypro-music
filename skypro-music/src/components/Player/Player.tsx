"use client";

import styles from "./Player.module.css";
import { useCallback, useEffect, useRef, useState } from "react";
import Icon from "../Icon/Icon";
import Image from "next/image";
import ProgressBar from "../ProgressBar/ProgressBar";
import { formatTime } from "@/utils/helpers";
import { useAppDispatch, useAppSelector } from "@/store/store";
import {
  setNextTrack,
  setPlayingState,
  togglePlaying,
  toggleShuffle,
  setPrevTrack,
  setPlaylistState,
  setTrackCurrentTime,
} from "@/store/features/trackSlice";
import { useLikeTrack } from "@/hooks/useLikeTracks";

function Player() {
  const { trackState, playingState, playlistState, shuffleActiveState } =
    useAppSelector((state) => state.track);

  const dispatch = useAppDispatch();

  const { isLiked, handleLike } = useLikeTrack({ track: trackState });

  const audioRef = useRef<HTMLAudioElement | null>(null);

  const [isRepeatActive, setIsRepeatActive] = useState(false);
  const [volume, setVolume] = useState(0.5);
  const [currentTime, setCurrentTime] = useState(0);

  const duration = audioRef.current?.duration || 0;

  const handleEnded = useCallback(() => {
    dispatch(setNextTrack());
    dispatch(setTrackCurrentTime(0));
  }, [dispatch]);

  useEffect(() => {
    const audio = audioRef.current;

    if (audio) {
      if (trackState) {
        audio.src = trackState.track_file;
      }

      audio.play();
      dispatch(setPlayingState(true));

      audio.addEventListener("ended", handleEnded);
    }

    return () => {
      if (audio) {
        audio.removeEventListener("ended", handleEnded);
      }
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [trackState, dispatch]);

  useEffect(() => {
    if (audioRef.current) {
      if (playingState) {
        audioRef.current.play();
      } else {
        audioRef.current.pause();
      }
    }
  }, [playingState]);

  useEffect(() => {
    audioRef.current!.volume = volume;
  }, [volume]);

  const togglePlay = () => {
    if (playingState) {
      audioRef.current!.pause();
    } else {
      audioRef.current!.play();
    }
    dispatch(togglePlaying());
  };

  const toggleRepeat = () => {
    setIsRepeatActive((prevState) => !prevState);
  };

  const toggleShuffleButton = () => {
    dispatch(toggleShuffle());
    dispatch(setPlaylistState({ tracks: playlistState }));
    dispatch(setNextTrack());
    dispatch(setTrackCurrentTime(0));
  };

  const handleButtonNextClick = () => {
    dispatch(setNextTrack());
    dispatch(setTrackCurrentTime(0));
  };

  const handleButtonPrevClick = () => {
    dispatch(setPrevTrack());
    dispatch(setTrackCurrentTime(0));
  };

  const handleLikeButton = async (event: React.MouseEvent<HTMLElement>) => {
    handleLike(event);
  };

  return (
    <div className={styles.bar} data-testid="player-container">
      <div className={styles.content}>
        <audio
          ref={audioRef}
          src={trackState?.track_file}
          loop={isRepeatActive ? true : false}
          onTimeUpdate={(e) => {
            setCurrentTime(e.currentTarget.currentTime);
            dispatch(setTrackCurrentTime(currentTime));
          }}
        />
        <div>
          <ProgressBar
            max={duration}
            value={currentTime}
            step={0.01}
            onChange={(e) => {
              audioRef.current!.currentTime = Number(e.target.value);
            }}
          ></ProgressBar>
        </div>

        <div className={styles.playerBlock}>
          <div className={styles.player}>
            <div className={styles.controls}>
              <Icon
                wrapperClass={styles.buttonPrev}
                iconClass={styles.buttonPrevSvg}
                name="icon-prev"
                onClick={handleButtonPrevClick}
              />
              {!playingState && (
                <Icon
                  data-testid="player-button"
                  wrapperClass={styles.buttonPlay}
                  iconClass={styles.buttonPlaySvg}
                  name="icon-play"
                  onClick={() => {
                    dispatch(setPlayingState(true));
                    audioRef.current!.play();
                  }}
                />
              )}
              {playingState && (
                <div
                  className={styles.buttonPause}
                  onClick={() => {
                    dispatch(setPlayingState(false));
                    audioRef.current!.pause();
                  }}
                >
                  <Image
                    src="/img/icon/pause.svg"
                    alt=""
                    width={15}
                    height={19}
                  />
                </div>
              )}

              <Icon
                wrapperClass={styles.buttonNext}
                iconClass={styles.buttonNextSvg}
                name="icon-next"
                onClick={handleButtonNextClick}
              />
              <Icon
                wrapperClass={styles.buttonRepeat}
                iconClass={
                  isRepeatActive
                    ? styles.buttonRepeatSvgActive
                    : styles.buttonRepeatSvg
                }
                name="icon-repeat"
                onClick={toggleRepeat}
              />
              <Icon
                wrapperClass={styles.buttonShuffle}
                iconClass={
                  shuffleActiveState
                    ? styles.buttonShuffleSvgActive
                    : styles.buttonShuffleSvg
                }
                name="icon-shuffle"
                onClick={toggleShuffleButton}
              />
            </div>

            <div className={styles.trackPlay}>
              <div className={styles.trackPlayContainer}>
                <Icon
                  wrapperClass={styles.trackPlayImage}
                  iconClass={styles.trackPlayImageSvg}
                  name="icon-note"
                />
                <div className={styles.trackPlayName}>
                  <a
                    className={styles.trackPlayNameLink}
                    href="#"
                    onClick={togglePlay}
                  >
                    {trackState?.name}
                  </a>
                </div>
                <div className={styles.trackPlayAuthor}>
                  <a className={styles.trackPlayAuthorLink} href="#">
                    {trackState?.author}
                  </a>
                </div>
              </div>

              <div className={styles.trackTime}>
                <p className={styles.trackTimeFigures}>
                  {formatTime(currentTime)}
                </p>{" "}
                /{" "}
                <p className={styles.trackTimeFigures}>
                  {formatTime(duration)}
                </p>
              </div>

              <div className={styles.trackPlayLikeDis}>
                <Icon
                  wrapperClass={styles.trackPlayLike}
                  iconClass={
                    isLiked
                      ? styles.trackPlayLikeSvgActive
                      : styles.trackPlayLikeSvg
                  }
                  name="icon-like"
                  onClick={handleLikeButton}
                />
              </div>
            </div>
          </div>

          <div className={styles.volumeBlock}>
            <div className={styles.volumeContent}>
              <Icon
                wrapperClass={styles.volumeImage}
                iconClass={styles.volumeSvg}
                name="icon-volume"
              />
              <div className={styles.volumeProgress}>
                <input
                  className={styles.volumeProgressLine}
                  type="range"
                  name="range"
                  min="0"
                  max="1"
                  step="0.01"
                  value={volume}
                  onChange={(e) => setVolume(Number(e.target.value))}
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
}

export default Player;
