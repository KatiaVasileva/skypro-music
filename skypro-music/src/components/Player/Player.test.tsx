import ReduxProvider from "@/store/ReduxProvider";
import { fireEvent, render, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import Player from "./Player";

describe("Player component", () => {
  let playStub: jest.SpyInstance<Promise<void>, [], any>;
  let pauseStub: jest.SpyInstance;

  beforeEach(() => {
    playStub = jest.spyOn(window.HTMLMediaElement.prototype, "play");
    pauseStub = jest
      .spyOn(window.HTMLMediaElement.prototype, "pause")
      .mockImplementation(() => {});
  });

  afterEach(() => {
    playStub.mockRestore();
    pauseStub.mockRestore();
  });

  it("renders correctly", () => {
    playStub.mockResolvedValue();

    const { container } = render(
      <ReduxProvider>
        <Player />
      </ReduxProvider>
    );
    expect(container).toMatchSnapshot();
  });
});
