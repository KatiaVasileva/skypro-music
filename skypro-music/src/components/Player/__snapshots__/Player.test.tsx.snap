// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player component renders correctly 1`] = `
<div>
  <div
    class="bar"
    data-testid="player-container"
  >
    <div
      class="content"
    >
      <audio />
      <div>
        <input
          class="progressBar"
          max="0"
          min="0"
          step="0.01"
          type="range"
          value="0"
        />
      </div>
      <div
        class="playerBlock"
      >
        <div
          class="player"
        >
          <div
            class="controls"
          >
            <div
              class="buttonPrev"
            >
              <svg
                class="buttonPrevSvg"
              >
                <use
                  xlink:href="/img/icon/sprite.svg#icon-prev"
                />
              </svg>
            </div>
            <div
              class="buttonPause"
            >
              <img
                alt=""
                data-nimg="1"
                decoding="async"
                height="19"
                loading="lazy"
                src="/img/icon/pause.svg"
                style="color: transparent;"
                width="15"
              />
            </div>
            <div
              class="buttonNext"
            >
              <svg
                class="buttonNextSvg"
              >
                <use
                  xlink:href="/img/icon/sprite.svg#icon-next"
                />
              </svg>
            </div>
            <div
              class="buttonRepeat"
            >
              <svg
                class="buttonRepeatSvg"
              >
                <use
                  xlink:href="/img/icon/sprite.svg#icon-repeat"
                />
              </svg>
            </div>
            <div
              class="buttonShuffle"
            >
              <svg
                class="buttonShuffleSvg"
              >
                <use
                  xlink:href="/img/icon/sprite.svg#icon-shuffle"
                />
              </svg>
            </div>
          </div>
          <div
            class="trackPlay"
          >
            <div
              class="trackPlayContainer"
            >
              <div
                class="trackPlayImage"
              >
                <svg
                  class="trackPlayImageSvg"
                >
                  <use
                    xlink:href="/img/icon/sprite.svg#icon-note"
                  />
                </svg>
              </div>
              <div
                class="trackPlayName"
              >
                <a
                  class="trackPlayNameLink"
                  href="#"
                />
              </div>
              <div
                class="trackPlayAuthor"
              >
                <a
                  class="trackPlayAuthorLink"
                  href="#"
                />
              </div>
            </div>
            <div
              class="trackTime"
            >
              <p
                class="trackTimeFigures"
              >
                0:00
              </p>
               
              /
               
              <p
                class="trackTimeFigures"
              >
                0:00
              </p>
            </div>
            <div
              class="trackPlayLikeDis"
            >
              <div
                class="trackPlayLike"
              >
                <svg
                  class="trackPlayLikeSvg"
                >
                  <use
                    xlink:href="/img/icon/sprite.svg#icon-like"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <div
          class="volumeBlock"
        >
          <div
            class="volumeContent"
          >
            <div
              class="volumeImage"
            >
              <svg
                class="volumeSvg"
              >
                <use
                  xlink:href="/img/icon/sprite.svg#icon-volume"
                />
              </svg>
            </div>
            <div
              class="volumeProgress"
            >
              <input
                class="volumeProgressLine"
                max="1"
                min="0"
                name="range"
                step="0.01"
                type="range"
                value="0.5"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
